classes/observation_manager.php

```diff
@@ -0,0 +1,222 @@
+     * @return mixed True if successful. If $returnid is True and $newinstance is True, returns ID
+     */
+    public static function modify_observation_point($data, bool $newinstance = false, bool $returnid = false,
+        string $tablename = 'observation_points') {
+
+        $data = (object)$data;
+
+        global $DB;
+
+        if (property_exists($data, 'max_grade')) {
+            // Ensure maxgrade (if set) is an int.
+            if (!is_int($data->max_grade)) {
+                throw new coding_exception("Property max_grade must be an int.");
+            }
+
+            // Ensure maxgrade (if set) is not negative.
+            if ($data->max_grade < 0) {
+                throw new coding_exception("Property max_grade cannot be negative");
+            }
+        }
+
+        if (property_exists($data, 'res_type')) {
+            // Get record, MUST_EXIST is passed so will except if res_type is invalid.
+            $DB->get_record('observation_res_type_map', ['res_type' => $data->res_type], '*', MUST_EXIST);
+        }
+
+        if ($newinstance) {
+            // Get the max observation point list_order to place this new one after it.
+            // Get all the points to get the bounds.
+            $allpoints = self::get_observation_points($data->obs_id);
+            $alllistorders = array_column($allpoints, 'list_order');
+
+            // Default to zero, update to max if a point already exists.
+            $maxordering = 0;
+            if (count($alllistorders) != 0) {
+                $maxordering = max($alllistorders);
+            }
+
+            // Update $data ordering property to be the max_ordering + 1 (place at end of list).
+            $data->list_order = $maxordering + 1;
+
+            // Insert.
+            return $DB->insert_record($tablename, $data, $returnid);
+        } else {
+            return $DB->update_record($tablename, $data);
+        }
+    }
+
+    /**
+     * Gets observation point data
+     * @param int $observationid ID of observation instance
+     * @param int $pointid ID of the observation point
+     * @param string $tablename database table name
+     * @return object existing point data
+     */
+    public static function get_existing_point_data(int $observationid, int $pointid,
+        string $tablename = 'observation_points'): object {
+
+        global $DB;
+        return $DB->get_record($tablename, ['id' => $pointid, 'obs_id' => $observationid], '*', MUST_EXIST);
+    }
+
+    /**
+     * Get all observation points in an observation instance
+     * @param int $observationid ID of the observation instance
+     * @param string $sortby column to sort by
+     * @param string $tablename database tablename
+     * @return array array of database objects obtained from database
+     */
+    public static function get_observation_points(int $observationid, string $sortby='list_order',
+        string $tablename='observation_points'): array {
```

Comment:
Can this be indented 1 more? 8 spaces vs 4 for the function body. Creates a visual difference between the function declaration body